fdbserver/storageserver.actor.cpp

```diff
@@ -1092,98 +1146,98 @@ ACTOR Future<Void> getValueQ( StorageServer* data, GetValueRequest req ) {
-					}
-					wait(watchFuture);
-					--data->numWatches;
-					data->watchBytes -= (req.key.expectedSize() + req.value.expectedSize() + WATCH_OVERHEAD_BYTES);
-				} catch( Error &e ) {
-					--data->numWatches;
-					data->watchBytes -= (req.key.expectedSize() + req.value.expectedSize() + WATCH_OVERHEAD_BYTES);
-					throw;
+			// TODO: Change the calculation of watchBytes to account for changes
+			data->watchBytes += (metadata->key.expectedSize() + metadata->value.expectedSize() + WATCH_OVERHEAD_BYTES);
+			try {
+				if(latest < minVersion) {
+					// If the version we read is less than minVersion, then we may fail to be notified of any changes that occur up to or including minVersion
+					// To prevent that, we'll check the key again once the version reaches our minVersion
+					watchFuture = watchFuture || data->version.whenAtLeast(minVersion);
 				}
-			} catch( Error &e ) {
-				if( e.code() != error_code_transaction_too_old ) {
-					throw;
+				if(BUGGIFY) {
+					// Simulate a trigger on the watch that results in the loop going around without the value changing
+					watchFuture = watchFuture || delay(deterministicRandom()->random01());
 				}
-
-				TEST(true); // Reading a watched key failed with transaction_too_old
+				wait(watchFuture);
+				// TODO: Change the calculation of watchBytes to account for changes
+				data->watchBytes -= (metadata->key.expectedSize() + metadata->value.expectedSize() + WATCH_OVERHEAD_BYTES);
+			} catch( Error &e ) {
+				// TODO: Change the calculation of watchBytes to account for changes
+				data->watchBytes -= (metadata->key.expectedSize() + metadata->value.expectedSize() + WATCH_OVERHEAD_BYTES);
+				throw;
+			}
+		} catch( Error &e ) {
+			if( e.code() != error_code_transaction_too_old ) {
+				throw e;
 			}
 
-			watchFuture = data->watches.onChange(req.key);
-			wait(data->version.whenAtLeast(data->data().latestVersion));
+			TEST(true); // Reading a watched key failed with transaction_too_old
 		}
-	} catch (Error& e) {
-		if(!canReplyWith(e))
-			throw;
-		data->sendErrorWithPenalty(req.reply, e, data->getPenalty());
+
+		watchFuture = data->watches.onChange(metadata->key);
+		wait(data->version.whenAtLeast(data->data().latestVersion));
+	}
+}
+
+void checkCancelWatchImpl( StorageServer* data, WatchValueRequest req ) {
+	Reference<ServerWatchMetadata> metadata = data->getWatchMetadata(req.key.contents());
+	if(metadata.isValid() && metadata->versionPromise.getFutureReferenceCount() == 1) {
+		// last watch timed out so cancel watch_impl and delete key from the map
+		data->deleteWatchMetadata(req.key.contents());
+		metadata->watch_impl.cancel();
 	}
-	return Void();
 }
 
-ACTOR Future<Void> watchValueQ( StorageServer* data, WatchValueRequest req ) {
-	state Span span("SS:watchValue"_loc, { req.spanContext });
-	state Future<Void> watch = watchValue_impl( data, req, span.context );
+ACTOR Future<Void> watchValueQ( StorageServer* data, WatchValueRequest req, Future<Version> resp ) {
```

Comment:
Add the `Span` back to this function, and set its parent to a new `SpanContext` param added to `watchValueQ` (you can set this param to `span.context` in `serveWatchValueRequestsImpl`).